src/main/flight/failsafe.c

```diff
@@ -115,58 +137,57 @@ static void failsafeOnValidDataReceived(void)
 
 void failsafeUpdateState(void)
 {
-    uint8_t i;
+    bool receivingRxData = failsafeIsReceivingRxData();
+    bool armed = ARMING_FLAG(ARMED);
 
-    if (!failsafeHasTimerElapsed()) {
-        return;
+    if (receivingRxData) {
+        failsafeState.phase = FAILSAFE_IDLE;
+        failsafeState.active = false;
     }
 
-    if (!failsafeIsEnabled()) {
-        failsafeReset();
-        return;
+    switch (failsafeState.phase) {
+        case FAILSAFE_IDLE:
+            if (!receivingRxData && armed) {
+                failsafeState.phase = FAILSAFE_RX_LOSS_DETECTED;
+            }
+            break;
+
+        case FAILSAFE_RX_LOSS_DETECTED:
+            if (failsafeShouldForceLanding(armed)) {
+                // Stabilize, and set Throttle to specified level
+                failsafeActivate();
+            }
+            break;
+
+        case FAILSAFE_LANDING:
+            if (armed) {
+                failsafeApplyControlInput();
+            }
+
+            if (failsafeShouldHaveCausedLandingByNow() || !armed) {
+
+                failsafeState.phase = FAILSAFE_LANDED;
+                failsafeState.active = false;
+                mwDisarm();
+            }
+            break;
```

Comment:
@ProDrone yes, i see there is quite a bit of overlap from what you worked on.  Don't worry though all your good stuff will make it in due course, this happens sometimes.   I'm very happy that you are following this since we now have even more developers and people looking at the code and discussing it, etc.

@ledvinap the race conditions are clearly shown in the unit tests in eb74735ee8a5a4c1761c0487af7237a9ce4045ab